state/accountsDB.go

```diff
@@ -1089,53 +1105,89 @@ func (adb *AccountsDB) SnapshotState(rootHash []byte) {
 	adb.mutOp.Lock()
 	defer adb.mutOp.Unlock()
 
-	trieStorageManager := adb.mainTrie.GetStorageManager()
-	epoch, err := trieStorageManager.GetLatestStorageEpoch()
-	if err != nil {
-		log.Error("snapshotState error", "err", err.Error())
+	trieStorageManager, epoch, shouldTakeSnapshot := adb.prepareSnapshot(rootHash)
+	if !shouldTakeSnapshot {
 		return
 	}
 
-	snapshotAlreadyTaken := bytes.Equal(adb.lastSnapshot.rootHash, rootHash) && adb.lastSnapshot.epoch == epoch
-	if !trieStorageManager.ShouldTakeSnapshot() || snapshotAlreadyTaken {
+	log.Info("starting snapshot user trie", "rootHash", rootHash, "epoch", epoch)
+	errChan := make(chan error, 1)
+	stats := newSnapshotStatistics(1)
+	go func() {
+		leavesChannel := make(chan core.KeyValueHolder, leavesChannelSize)
+		stats.NewSnapshotStarted()
+		trieStorageManager.TakeSnapshot(rootHash, rootHash, leavesChannel, errChan, stats, epoch)
+		adb.snapshotUserAccountDataTrie(true, rootHash, leavesChannel, errChan, stats, epoch)
+		trieStorageManager.ExitPruningBufferingMode()
+
+		stats.wg.Done()
+	}()
+
+	go adb.markActiveDBAfterSnapshot(stats, errChan, rootHash, "snapshotState user trie", epoch)
+
+	adb.waitForCompletionIfRunningInImportDB(stats)
```

Comment:
Perhaps, in the future, we can also `waitForCompletionIfHeavilySyncing()` (when out of sync). From my understanding of the logs (if correct), that would greatly reduce the snapshot time. Does this make sense?

Reply:
E.g. import db finishes snapshot in 2-3 minutes.

Reply:
The snapshot process is started on a goroutine. During import-db, epochs are processed much faster than an actual epoch time, so there is a chance that an epoch is processed faster than it will take the snapshot to finish. `adb.waitForCompletionIfRunningInImportDB(stats)` is added to assure that the processing will wait for the snapshot to finish.